lib/bridge.js

```diff
@@ -773,52 +783,47 @@ Bridge.prototype._onEvent = function(event) {
         }
     }
 
-    var request = this._requestFactory.newRequest({ data: event });
-    var context = new BridgeContext({
-        sender: event.user_id,
-        target: event.state_key,
-        room: event.room_id
-    });
-    var data = {
+    const request = this._requestFactory.newRequest({ data: event });
+    const contextReady = this._getBridgeContext(event);
+    const dataReady = contextReady.then(c => ({
         request: request,
-        context: context
-    };
-
-    var promise;
-    if (this.opts.disableContext) {
-        promise = Promise.resolve();
-        data.context = null;
-    }
-    else {
-        promise = context.get(this._roomStore, this._userStore);
-    }
+        context: c
+    }));
 
     if (this.opts.queue.perRequest) {
-        promise = Promise.settle([
-            promise,
-            this._prevRequestPromise
-        ]);
+        dataReady = prolong(dataReady, this._prevRequestPromise)
```

Comment:
Out of interest, is there anytthing wrong with `dataReady = this._prevRequestPromise.finally(() => dataReady)`?

`finally` will fire no matter the outcome, and if the promise has already resolved/rejected, it will still fire.

Reply:
I am new to the JS Promise world. To me it looks like some flattening magic is going on in your example code, but I think it does the right thing :)